### PR TITLE
Support microservice template

### PR DESCRIPTION
**Pull request by carmenlau**
- Download template 
- Combine user code and template files when archiving user code

### CODE REVIEW

File: src/commands/app/deploy.ts [hunk 1]
```diff
@@ -54,7 +49,7 @@ async function archiveMicroserviceSrc(
     ) => Promise<string[]>;
   } = {
     '.dockerignore': dockerignorePaths,
-    '.skyignore': skyignorePaths
+    '.skyignore': dockerignorePaths
```
Comment (louischan-oursky) [inline src/commands/app/deploy.ts]: Since we always use `dockerignorePaths` now, `ignoreFuncMap` is now redundant.

File: src/ignore.ts [hunk 2]
```diff
@@ -5,7 +5,6 @@ import {
 } from 'fs';
 import { promisify } from 'util';
 import { join, relative } from 'path';
-import gitignore from 'ignore';
```
Comment (louischan-oursky) [inline src/ignore.ts:8]: Can we also remove the dependency on `ignore`?

File: src/commands/app/deploy.ts (hunk 1 above)
Comment (carmenlau) [reply src/commands/app/deploy.ts]: `ignoreFuncMap` was removed in commit `0770f51`

**Comment by carmenlau** [comment]
Updated!:)

File: src/commands/app/deploy.ts [hunk 3]
```diff
@@ -29,53 +32,121 @@ function createArchiveReadStream(archivePath: string) {
...
+): Promise<void> {
+  return new Promise((resolve, _) => {
+    const z = zlib.createGzip();
+    const folders = Object.keys(folderToPathsMap);
+    createPack(folders, folderToPathsMap, null, 0)
+      .pipe(z)
+      .pipe(fs.createWriteStream(archivePath))
+      .on('finish', function() {
+        resolve();
+      });
```
Comment (kiootic) [inline src/commands/app/deploy.ts]: We should handle `error` event and reject the promise.

File: src/commands/app/deploy.ts [hunk 4]
```diff
@@ -29,53 +32,121 @@ function createArchiveReadStream(archivePath: string) {
...
+  });
+}
+
+function createPack(
+  folders: string[],
+  folderToPathsMap: { [folder: string]: string[] },
+  pack?: any,
+  cur: number = 0
+): any {
```
Comment (kiootic) [inline src/commands/app/deploy.ts:135]: We should install `@types/tar-fs` and use better typings.

File: src/commands/app/deploy.ts [hunk 5]
```diff
@@ -217,6 +292,50 @@ async function confirmIfItemsWillBeRemovedInNewDeployment(
...
+    'skygear-templates',
+    encodeURIComponent(templateName)
+  );
+}
+
+async function downloadTemplateIfNeeded(
+  deployments: DeploymentItemsMap,
+  cacheTemplate: boolean = false
+) {
+  const templatesToDownloadSet: { [name: string]: any } = {};
```
Comment (kiootic) [inline src/commands/app/deploy.ts]: `const templatesToDownloadSet = new Set<string>();`?

File: src/commands/app/deploy.ts (hunk 4 above)
Comment (carmenlau) [reply src/commands/app/deploy.ts:135]: `@types/tar-fs` is not up to date, `PackOptions` doesn't have the `pack` attribute :(
Do you think we should define type for `tar-fs` ourselves?

Comment (kiootic) [reply src/commands/app/deploy.ts:135]: Seems we're using just a small part of `tar-fs` API. I think it is doable to just define the types of API we used.

Comment (carmenlau) [reply src/commands/app/deploy.ts:135]: sure:)

**Review by louischan-oursky — APPROVED** [review]

**Review by kiootic — APPROVED** [review]

**Pull request merged**